### PR TITLE
Do not set the dockpanel as parent of the tabbar

### PR DESCRIPTION
**Pull request by brichet**
This PR fixes https://github.com/jupyterlab/lumino/issues/604

~~It adds an attribute to Widget to disable the tracking from `focusTracker`, event if its elements can get the focus.~~

~~By default all the widgets are trackable, except for the `TabBar`.~~

It avoids setting the parent of the tab bar in the dock panel.

### CODE REVIEW

**Comment by fcollonval** [comment]
Thanks a lot for digging this one up @brichet

I would rather avoid adding a new attribute for this.

Looking at the current behavior, when the dock panel creates a split, it adds a tab bar. When that tab bar is attached to the panel, its `parent` is set.

https://github.com/jupyterlab/lumino/blob/07d4bab5b73b753332d941495c27109fea9d25c2/packages/widgets/src/docklayout.ts#L1133

That triggers the `child-added` / `child-removed` message listened by the focus tracker.

https://github.com/jupyterlab/lumino/blob/07d4bab5b73b753332d941495c27109fea9d25c2/packages/widgets/src/widget.ts#L232

The addition of the split handle does not triggered such messages as they are directly attached to the HTML element.

---

This is actually wrong as adding/removing a tab bar should be transparent for dock panels. The message `child-added` and `child-removed` should only be triggered for child panels.

The wrong behavior can also be seen in JupyterLab by the wrongly addition of the class `jp-Activity` to the tab bars (it should only be on the `MainAreaWidget`) :

tabbar-issue

---

Therefore I think a better fix would be to not set the tab bar parent, simply attaching it as it is done for the handle.

cc @afshin @krassowski what do you think?

> Another approach could be the inhibition of the `child-added`/`child-removed` message emission. But I don't see a good way of doing this without overriding the `set parent` method of the tab bar. But then it may have unwanted effect for downstream code consuming tab bar outside a dock panel (like the weird usage in the JupyterLab sidebars).

**Comment by brichet** [comment]
> Therefore I think a better fix would be to not set the tab bar parent, simply attaching it as it is done for the handle.

Thanks for the suggestion @fcollonval, I reverted the previous changes and add a commit to avoid setting the parent of the tab bar.

File: packages/widgets/tests/src/dockpanel.spec.ts [hunk 1]
```diff
@@ -44,6 +44,15 @@ describe('@lumino/widgets', () => {
         let panel = new DockPanel();
         expect(panel.hasClass('lm-DockPanel')).to.equal(true);
       });
+
+      it('should not have tabbar as child', () => {
+        let panel = new DockPanel();
+        let w1 = new Widget();
+        panel.addWidget(w1);
```
[the remark this answers is not in the document]
Comment (brichet) [reply packages/widgets/tests/src/dockpanel.spec.ts:51]: Prior to this modification, the `TabBar` was added as a child widget of the `DockPanel`, as well as the document panels. The `TabBar` was only added if at least one widget had been added to the `DockPanel`.
This test adds a widget only to ensure the `TabBar` is not added to the child widget as well (which was the case prior to this PR).
In fact I believe it should be simplified, maybe this is what confused you.

```suggestion
        panel.addWidget(new Widget());
```

[the remark this answers is not in the document]
Comment (brichet) [reply packages/widgets/tests/src/dockpanel.spec.ts:51]: > Prior to this modification, the `TabBar` was added as a child widget of the `DockPanel`, as well as the document panels. The `TabBar` was only added if at least one widget had been added to the `DockPanel`. This test adds a widget only to ensure the `TabBar` is not added to the child widgets list (which was the case prior to this PR). In fact I believe it should be simplified, maybe this is what confused you.

[the remark this answers is not in the document]
Comment (gabalafou) [reply packages/widgets/tests/src/dockpanel.spec.ts:51]: > The TabBar was only added if at least one widget had been added to the DockPanel.

Without knowing these implementation details, somebody reading this test has to guess at the intent of the `panel.addWidget()` call. Could you clarify your intent with a comment?

[the remark this answers is not in the document]
Comment (brichet) [reply packages/widgets/tests/src/dockpanel.spec.ts]: Done in https://github.com/jupyterlab/lumino/pull/606/commits/f0a3d334983c5a31a6c52a8433293cee64c0f9f2

**Review by fcollonval — APPROVED** [review]
Thanks @brichet

**Pull request merged**